tests: kernel: pipe: testcases to improve code coverage for pipes

s-kelvin opened the pull request:
Added back test_pipe_alloc because the z_thread_malloc called in the API has been updated to use k_heap instead of k_mem_pool.
Adjusted test_resource_pool_auto_free by replacing z_mem_pool_malloc with k_heap_alloc. 
Added new test_k_pipe_cleanup to cover one more branch in k_pipe_cleanup. 
Modified test_half_pipe_put_get to cover branches for (reader != NULL) in k_pipe_put. 
Added test test_pipe_get_put to cover branches for (writer != NULL) in k_pipe_get. 
Added other trivial tests to cover branches of input validity check.
The rest of the original testcases (deleted by Andy) are discarded due to their dependence on k_pipe_block_put, which is also going to be deprecated per [31212](https://github.com/zephyrproject-rtos/zephyr/pull/31212). 

Line coverage has been improved from 37.4% to 89.6%. Function coverage has been improved from 43.8% to 100%. Branch coverage has been improved from 19.8% to 65.6%. Will continue to work on branch coverage.

LeiW000 commented:
@s-kelvin , have you checked "buildbite/zephyr", which failed?

LeiW000 commented:
@andyross , could you take look at it? Kelvin is adding back the test case codes, which you removed before.

dcpleung commented on tests/kernel/pipe/pipe_api/src/main.c:
```diff
@@ -13,18 +13,27 @@
 extern void test_pipe_block_put_sema(void);
 extern void test_pipe_get_put(void);
-extern void test_half_pipe_get_put(void);
+
+extern void test_pipe_get_put(void);
+
```
Duplicate.

s-kelvin replied:
Thanks. Fixed

dcpleung commented on tests/kernel/pipe/pipe_api/prj.conf:
```diff
@@ -3,3 +3,4 @@ CONFIG_IRQ_OFFLOAD=y
 CONFIG_TEST_USERSPACE=y
 CONFIG_DYNAMIC_OBJECTS=y
 CONFIG_MP_NUM_CPUS=1
+CONFIG_USERSPACE=y
```
Don't force userspace. This test suite needs to also work without userspace.

s-kelvin replied:
I add one more test in yaml which enables userspace instead.

LeiW000 commented:
Can anyone help to merge it? The two reviewers already approved it.

MaureenHelm commented:
@andyross can you review please?

nashif commented on tests/kernel/pipe/pipe_api/testcase.yaml:
```diff
@@ -1,3 +1,7 @@
   kernel.pipe.api:
-      tags: kernel userspace
+    tags: kernel userspace
+  kernel.pipe.api.userspace:
```
I am failing to see why you need this scenario with userspace enabled, all tests should be able to run on both userspace and non-userspace systems with the same code.

nashif replied:
This btw, will have the same test run twice the same way on platforms which enable userspace by default.

s-kelvin replied:
Thanks for pointing out. I got confused before. Removed the extra test.

nashif commented on tests/kernel/pipe/pipe_api/src/test_pipe_contexts.c:
```diff
@@ -189,27 +301,85 @@ void test_pipe_user_thread2thread(void)
 
+/**
+ * @brief Test resource pool free
+ * @see z_mem_pool_malloc()
```
we do not have z_mem_pool_malloc in the tree.

s-kelvin replied:
Replaced the comment by k_heap_alloc.

nashif reviewed (commented):
I do not see z_mem APIs being replaced here, looks like the commit message and PR title are outdated.

SecureSheII committed:
tests: kernel: pipe: testcases to improve code coverage for pipes

Added back test_pipe_alloc because the z_thread_malloc called in the
API has been updated to use k_heap instead of k_mem_pool.
Adjusted test_resource_pool_auto_free by replacing z_mem_pool_malloc
with k_heap_alloc. Added new test_k_pipe_cleanup to cover one more
branch in k_pipe_cleanup. Modified test_half_pipe_put_get to cover
branches for (reader != NULL) in k_pipe_put. Added test
test_pipe_get_put to cover branches for (writer != NULL)
in k_pipe_get. Added trivial tests to cover input validity checks.
Line coverage has been improved by 52%, function cov by 56%, and
branch cov by 46%.

Signed-off-by: Shihao Shen <shihao.shen@intel.com>

s-kelvin commented:
> I do not see z_mem APIs being replaced here, looks like the commit message and PR title are outdated.

@nashif z_mem APIs are replaced by k_heap APIs for those testcases that were previously deleted by Andy. This commit includes those replacements by adding back these testcases after replacing their dependence on z_mem. Also this commit includes other new tests developed to improve the code coverage. But yes the commit message is a little confusing so I've changed the title. Hope it makes sense now.
